freebsd/FreeBSDProcess.c

```diff
@@ -73,22 +74,40 @@ void Process_delete(Object* cast) {
    free(this);
 }
 
+static_assert(MAX_SCHEDCLASS == 5, "The lookup table below isn't in sync with the SCHEDCLASS_ enum. Please update.");
+static const char FreeBSD_schedclassChars[MAX_SCHEDCLASS] = {
```

Comment:
1. The static assert is probably unnecessary.
2. It's good to give some rationale on how the characters are chosen this way, because this thing is likely added into htop documentation. E.g. is the blank character (`SCHEDCLASS_TIMESHARE`) considered normal for a FreeBSD process, and why is `SCHEDCLASS_ITHD` represented as a hyphen?

Reply:
I've put in the static assert for a few reasons:
- leave a compiler hint whenever someone changes the schedclass enum
- and thus prevent access to uninitialised data when such changes happen

Yes, it is unlikely this enum will ever change - this is just being less hostile to whoever works on this code and to prevent potential bugs.

Reply:
@herrhotzenplotz There are other mapping tables from an enum in htop codebase like this, and none of them needs a static assert. It's unlikely you need one here either.
When an array is initialized with a "named index" initializer like this, unspecified indices would have zero value. If you need an assert, do so on the mapping side so that the mapped character isn't `'\0'`.

Reply:
> There are other mapping tables from an enum in htop codebase like this, and none of them needs a static assert.

Well I'd say this would be a problem with those "other mapping tables". Having the static_assert in place does no harm and only improves things. I don't see why anyone would willingly remove this simple assurance.

> If you need an assert, do so on the mapping side so that the mapped character isn't '\0'.

I can certainly add this runtime assertion.

Reply:
> Well I'd say this would be a problem with those "other mapping tables". Having the static_assert in place does no harm and only improves things. I don't see why anyone would willingly remove this simple assurance.

First, it's not my decision, I only request this for the sake of code consistency. Second, having

Reply:
> Well I'd say this would be a problem with those "other mapping tables". Having the static_assert in place does no harm and only improves things. I don't see why anyone would willingly remove this simple assurance.

First, it's not my decision, I only request this for the sake of code consistency. Second, having an assertion of `MAX_SCHEDCLASS == 5` isn't helpful and it partly defeats the purpose of defining the `MAX_SCHEDCLASS` enum value (which is meant to increase over time).

Reply:
> First, it's not my decision, I only request this for the sake of code consistency.

Sacrificing being prone to bugs for code consistency seems like a very bad thing to me.

> Second, having an assertion of MAX_SCHEDCLASS == 5 isn't helpful

Yes it is helpful. The point is to stop compilation, let the developer update the lookup table, correct the static assertion and continue. This is what the assertion message tells. This is a pretty common way to guide someone making changes in large C codebases.

>  and it partly defeats the purpose of defining the MAX_SCHEDCLASS enum value (which is meant to increase over time).

The reason to define `MAX_SCHEDCLASS` is to define the static array of the correct size and to range-check incoming enum values.

Reply:
@herrhotzenplotz Would you raise this issue to a separate discussion?
IMO, a long term solution is for compiler to introduce an array/struct attribute to reject any implicit initialization of array/struct members.

Reply:
Sure!

Reply:
While I personally don't have any objections against this `static_assert` there are two reasons against this in this particular case:
- The `htop` source is targeted to be compile-able by any compliant C99 compiler. As [`static_assert`](https://en.cppreference.com/w/c/language/_Static_assert) was not introduced before C11 (as `_Static_assert`) or C23 proper this violates the language standard we are aiming for.
- The rest of the code base merely assumes for indexed initializers that if there's a `MAX_FOO` or `NUM_FOO` constant in an enum it's the last one (i.e. the one with the highest value). Thus having entries defined implicitly below that maximum value is totally fine and expected. The most prominent example is the set of internal IDs for columns.

If you really want to check, I'm totally fine with the runtime check as suggested by @Explorer09.

Reply:
@BenBE I am fully aware of the fact that C99 does *not* have `static_assert`. However, your code base has got compatibility code: https://github.com/htop-dev/htop/blob/4c0e96592d5f5f884aab9b29a3300ae5e1b6009b/Compat.h#L71-L73

In fact, this static_assert macro was used during my testing.

Reply:
Could you update this please to align with the rest of the code base?

Reply:
> Could you update this please to align with the rest of the code base?

I have done this weeks ago and have no idea why github still shows the static assertion here. It's gone in the full diff and I have converted them to runtime assertions instead as you suggested.